### ISSUE TITLE
Type compliance vs interface compliance, with substitution principle as the background

### ISSUE THREAD

**Issue by Hibou57** [microsoft/pyright#11788]
**Describe the bug**

A method is overridden in a derived class, in a way which makes it of a different type, but in a way which makes it fulfill the same interface. PyRight see an error. In a way, it is correct and understandable, while in another way, there is no need to complain, because it is correct for the interface.

There is no urgency about it, because the method can be made an instance method, just with an unused `self`, which just have to be written in way which please linters.

**Code or Screenshots**

```python
class A:
    def m(self):
        pass

class B(A):
    @staticmethod
    def m():
        pass
```

PyRight message is:
```text
  ... - error: Method "m" overrides class "A" in an incompatible manner
    Base method is declared as an instance method but override is not
    Positional parameter count mismatch; base method has 1, but override has 0
    Parameter 1 mismatch: base parameter "self" is keyword parameter, override parameter is position-only (reportIncompatibleMethodOverride)
```
Yes, `B.m` does not have the same type anymore, but from `A`’s point of view, `B.m` can be invoked the exact same way as `A.m`. To me (may be wrong) `B.m` does not break anything. It is just that it is implemented in such a way it does not need `self`, it is not that it prevent to invoke it from an instance (something with a self), it is just that it does not need it. Since it does not need it, it can be declared as a static method, which allows to call it from the class, it does not disallow to call it from an object. As I see it, `B.m` is more general than `A.m` so it can take place where `A.m` can take place (I’m not sure of the proper wording to use in terms of substitution). But this is indeed wrong if there exist any context where `B.m` would break anything, only due to its declaration (if its implementation may be wrong, this is another topic). Do such a context can exist? If yes, then there is no issue, it no, then PyRight is correct from some formal point of view, but not from a semantic point of view.

P.S. Happy thanks for the care given to issue https://github.com/microsoft/pyright/issues/11576

**Comment by rchiodo** [comment]
Thanks for the report. You are right that both `A().m()` and `B().m()` work. However, calling through an instance is only one part of the interface: Python also allows an instance method to be accessed through the class, with the instance supplied explicitly.

Using `A` and `B` from your example:

```python
def invoke(obj: A) -> None:
    type(obj).m(obj)

invoke(A())  # OK
invoke(B())  # TypeError: B.m() takes 0 positional arguments but 1 was given
```

For `A`, accessing `m` through the class produces a function that expects an explicit `self`. For `B`, it produces the static function, which accepts no arguments. The override therefore preserves `obj.m()` but breaks another supported operation. This is a concrete substitution failure, rather than just a difference in how the methods are declared.

The typing spec's [callable assignability rules](https://typing.python.org/en/latest/spec/callables.html#assignability-rules-for-callables) require a replacement callable to accept every argument combination accepted by the original. Its [`type[]` covariance rule](https://typing.python.org/en/latest/spec/special-types.html#type) also allows `type[B]` where `type[A]` is expected. These principles explain the compatibility concern here; they should not be read as an explicit specification of every aspect of Pyright's method-kind checks.

The recommendation is to keep the override as an instance method, even if its implementation does not use `self`:

```python
class B(A):
    def m(self):
        pass
```

Whether the implementation uses `self` is separate from whether its interface must accept it. Closing this as expected behavior, since the reported override does not preserve that interface.

**Comment by Hibou57** [comment]
Received soft and clear, thanks the enlightenment (will try to not forget it).
